h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1748,100 +1754,120 @@ private int getMovePriority(Chunk chunk) {
-                    break;
-                }
-                moveChunk(chunk, false);
-                iter.remove();
-            }
 
+            long leftmostBlock = Long.MAX_VALUE;
+            for (Chunk chunk : move) {
+                leftmostBlock = Math.min(leftmostBlock, chunk.block);
+            }
+            long sourceAreaBegin = leftmostBlock * BLOCK_SIZE;
+            long originalFileSize = getFileLengthInUse();
+            long originalBlokCount = (originalFileSize + BLOCK_SIZE - 1) / BLOCK_SIZE;
+            // we need to ensure that chunks moved within the following loop
+            // do not overlap with space just released by chunks moved before them,
+            // hence the need to reserve this area [sourceAreaBegin, originalFileSize)
             for (Chunk chunk : move) {
-                moveChunk(chunk, true);
+                moveChunk(chunk, sourceAreaBegin, originalFileSize);
             }
-
-            // update the metadata (store at the end of the file)
-            reuseSpace = false;
-            commit();
+            // update the metadata (hopefully within the file)
+            store(sourceAreaBegin, originalFileSize);
             sync();
 
-            Chunk chunk = lastChunk;
+            Chunk chunkToMove = lastChunk;
+            long biggestFileSize = getFileLengthInUse();
 
-            // now re-use the empty space
-            reuseSpace = true;
+            boolean movedToEOF = false;
+            // move all chunks, which previously did not fit before reserved area
+            // now we can re-use previously reserved area [sourceAreaBegin, originalFileSize),
+            // but need to reserve [originalFileSize, biggestFileSize)
             for (Chunk c : move) {
-                moveChunk(c, false);
+                if (c.block >= originalBlokCount) {
+                    moveChunk(c, originalFileSize, biggestFileSize);
+                    assert c.block < originalBlokCount;
+                    movedToEOF = true;
+                }
             }
-
-            // update the metadata (within the file)
-            commit();
-            sync();
-            if (moveChunk(chunk, false)) {
-                commit();
+            if ((movedToEOF || chunkToMove.block >= originalBlokCount) &&
+                    moveChunk(chunkToMove, originalFileSize, biggestFileSize)) {
+                // store a new chunk with updated metadata (hopefully within a file)
+                store(originalFileSize, biggestFileSize);
+                sync();
+                // and if last chunk did not fit within a file, since we can use
+                // previously reserved area [originalFileSize, biggestFileSize) now,
+                // try to move it closer to BOF
+                boolean moved = moveChunkInside(chunkToMove, biggestFileSize);
+                if (moveChunkInside(lastChunk, biggestFileSize) || moved) {
+                    store(0, 0);
```

Comment:
Here also need to ensure that the location of the last two chunks isn't overwrited by the new chunk created in line 1805.

Reply:
Fixed now, I think